### ISSUE TITLE
Denied Access On First Sign In

### ISSUE THREAD

**Issue by trey-wallis** [byuweb/byu-browser-oauth-implicit#15]
Hello,

I am a student developer at the college of mathematical and physical sciences. We are currently developing a web app for our upcoming student research conference using React and Redux. We access the CAS to let users login using their net id. We've noticed some odd behavior with logging in. Using the byu-user-info-oauth sign in button, we click on that link and get redirected to the login page for CAS. After entering in credentials and submitting and validating the duo request, we get redirected back to the website but the callback throws an error. It is only until we click sign again that we are authenticated as a user. We have checked our backend code and we pretty sure that it is not our React or Redux setup.

Way to replicate:
#1 Import the module in a react component
#2 Initialize the authentication observer
#3 Make a request using the byu-user-info-oauth button

Here are the error logs for the first sign in click, after entering in my credentials and being redirected back to our site.

![code](https://user-images.githubusercontent.com/40307803/73712179-89f19000-46c6-11ea-8edd-a5b55ad7e7cc.png)
![code2](https://user-images.githubusercontent.com/40307803/73712181-89f19000-46c6-11ea-8a25-7ea639a85b9f.png)

<br><br>
Here is the console output of the second click of the byu-user-info-oauth link. The page refreshes immediately and gives us our authenticated user token without asking us again to sign in
<br><br>

![authentication1](https://user-images.githubusercontent.com/40307803/73712422-44819280-46c7-11ea-8a75-089e0af1318b.png)
![authentication2](https://user-images.githubusercontent.com/40307803/73712425-45b2bf80-46c7-11ea-8d9e-7f6082884c03.png)

I can attach some sample code if needed.

Regards,
CPMS.

**Comment by trey-wallis** [comment]
In trying to duplicate the behavior again, I have noticed that it normally happens if I try signing into another browser through incognito mode when I'm currently logged in on another browser

**Comment by snelg** [comment]
Huh, yeah looks like LocalStorage in Firefox has some bugs/quirks, especially in Incognito mode.
It's wonky enough that they're completely rebuilding LocalStorage from the ground up. The "NextGen" version of LocalStorage is in pre-release.
I was able to reproduce your bug, but when I switched to that NextGen version the bug went away.
Toggle the `about:config` setting `dom.storage.next_gen` to `true` if you want to try it yourself.

Obviously I'm not necessarily saying you should immediately make that switch permanent, since Mozilla hasn't pushed it to final yet.

**Comment by snelg** [comment]
The actual data that's being stored in LocalStorage (and sometimes lost in Incognito mode) is relatively tiny. I'll see if we can switch that particular "save state" function to just use a Cookie instead of LocalStorage.

**Comment by snelg** [comment]
@trey-wallis Try using this version and see if it A) fixes your issue, and B) doesn’t cause any other issues:

`https://cdn.byu.edu/browser-oauth-implicit/experimental/init-cookie/implicit-grant.min.js`

i.e. wherever you have that reference in your own code, change `latest` to `experimental/init-cookie`

**Comment by snelg** [comment]
Tested as thoroughly as I can with new version, so rolled out. You can stay on (or switch back to) `latest`.

**Comment by trey-wallis** [comment]
Did you update the NPM module on https://www.npmjs.com/package/@byuweb/browser-oauth?

**Comment by snelg** [comment]
> Did you update the NPM module on https://www.npmjs.com/package/@byuweb/browser-oauth?

No changes to that package. That's the generic wrapper library. This issue was an implementation detail specific to this "Implicit Grant" provider.

**Comment by trey-wallis** [comment]
Gotcha. I'll update my packages and then try it

**Comment by trey-wallis** [comment]
I tried testing the sign in functionality on our application and the issue of the first login being declined has gone away. The API successfully redirects from CAS to our application as logged in on every first login. However, the issue still occurs on an incognito window. The first login attempt on an incognito window fails if we've already logged in on another browser. This behavior also continues if I log out of the first browser and then try logging in again on a different incognito window. 

---
Actually the behavior persists for all incognito windows on Firefox.
----

This doesn't really matter to me at this point, but I don't know if it's the desired behavior that you want.
